include/string_utils.h

```diff
@@ -380,15 +382,63 @@ std::optional<float> parse_percentage(const std::string_view s);
 //   else
 //       log_warning("%s was invalid", s.c_str());
 //
-std::optional<float> parse_prefixed_value(const char prefix, const std::string &s,
+std::optional<float> parse_prefixed_value(const char prefix, const std::string& s,
                                           const float min_value,
                                           const float max_value);
 
 // parse_prefixed_value clamped between 0 and 100
-std::optional<float> parse_prefixed_percentage(const char prefix, const std::string &s);
+std::optional<float> parse_prefixed_percentage(const char prefix,
+                                               const std::string& s);
 
 // tries to convert string to integer,
 // returns value only if succeeded
 std::optional<int> to_int(const std::string& value);
 
+#if defined(__GNUC__) || defined(__clang__)
+// Disable generic "format string is not a string literal (potentially
+// insecure)" warning on GCC/Clang. Of course, it's not a security issue for
+// us; we know what we're doing.
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wformat-security"
+#endif
+
+template <typename... Args>
+std::string format_string(const std::string& format, Args&&... args) noexcept
+{
+	// Perform a non-writing format to determine the size
+	const auto required_size = std::snprintf(nullptr,
+	                                         0,
+	                                         format.c_str(),
+	                                         std::forward<Args>(args)...);
+	if (required_size <= 0) {
+		return {};
+	}
+
+	// snprintf's length parameter specifies the maximum number of
+	// characters to be written without the trailing null. However, it still
+	// writes the trailing null into the buffer, so we need to include that
+	// in our allocation.
+	std::string result(static_cast<size_t>(required_size) +
+	                           static_cast<size_t>(1),
+	                   '\0');
+
+	std::snprintf(const_cast<char*>(result.c_str()),
```

Comment:
You can get the string's writeable and contiguous memory with the `.data()` call, which avoids need for `const_cast<char*>`.

Reply:
I'll check again but maybe not an option here. I learned that data() does not give you access to a null terminated C string; it's just the pure string data! Yes, sometime there will be a zero at the end by fluke, however that's not to be relied upon and thus is highly unsafe.

Reply:
oh, .data() /should/ be OK here though, because `result`'s entire length has been filled with the `'\0'` terminator in the constructor call.

Reply:
Yes, tests are still passing.